lk2nd/hw/gpio/gpio.c

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/* Copyright (c) 2022 Nikita Travkin <nikita@trvn.ru> */
+
+#include <debug.h>
+#include <stdint.h>
+#include <string.h>
+#include <libfdt.h>
+
+#include <lk2nd/gpio.h>
+
+#include "supplier.h"
+
+#define LK2ND_GPIOL_PROP_LEN 32
+
+struct gpio_desc gpiol_get(const void *dtb, int node, const char *name, uint32_t flags)
+{
+	char prop_name[LK2ND_GPIOL_PROP_LEN + sizeof("-gpios")] = "gpios";
+	const fdt32_t *prop_val;
+	struct gpio_desc pin = {0};
+	uint32_t config;
+	int len, ret;
+
+	if (name) {
+		strncpy(prop_name, name, LK2ND_GPIOL_PROP_LEN);
+		strcat(prop_name, "-gpios");
+	}
+	prop_val = fdt_getprop(dtb, node, prop_name, &len);
+
+	if (len < 0) {
+		dprintf(CRITICAL, "gpiol_get failed: Unable to get property: %d\n", len);
+		return gpiol_error(len);
+	} else if (len != 3 * sizeof(*prop_val)) {
+		dprintf(CRITICAL, "gpiol_get failed: gpios arrays are not supported. (len=%d)\n", len);
+		return gpiol_error(-1);
+	}
+
+	pin.pin = fdt32_to_cpu(prop_val[1]);
+	pin.dev = fdt32_to_cpu(prop_val[0]) & 0xff;
+	config = fdt32_to_cpu(prop_val[2]);
+
+	ret = gpiol_set_config(&pin, config);
+	if (ret < 0) {
+		dprintf(CRITICAL, "gpiol_get failed: Unable to configure gpio: %d\n", len);
+		return gpiol_error(ret);
+	}
+
+	if (flags & GPIOL_FLAGS_IN)
+		ret = gpiol_direction_input(pin);
+	else if (flags & GPIOL_FLAGS_OUT)
+		ret = gpiol_direction_output(pin, !!(flags & GPIOL_FLAGS_ASSERTED));
```

Comment:
This needs to be done inside the gpiol_set_config() so the pins are immediately correctly configured. For pm8x41 this can be just passed directly to pm8x41_gpio_config() (a bit confusingly it's `config->function` = 0/1 for low/high). For TLMM it should run `lk2nd_gpio_tlmm_set()` first, I guess